Ninja builds broken by circular dependencies

Unlike make, ninja doesn't handle well circular dependencies introduced by de7e0f565e358d67105ef0b8144b0deed7f549ec. Builds for pkgj are broken without patching vita.cmake like this 

```diff
--- vitasdk/share/vita.cmake	2020-01-09 14:10:23.000000000 -0500
+++ vitasdk2/share/vita.cmake	2020-01-09 21:22:57.699001000 -0500
@@ -84,12 +84,12 @@
   )

   ## SELF target
-  add_custom_target(${target}
+  add_custom_target(${target}.dep
     ALL
     DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target}
   )
   if(TARGET ${source})
-    add_dependencies(${target} ${source})
+    add_dependencies(${target}.dep ${source})
   endif()
 endmacro(vita_create_self)
 ##################################################
@@ -239,12 +239,12 @@
   )

   ## VPK target
-  add_custom_target(${target}
+  add_custom_target(${target}.dep
     ALL
     DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target}
   )
   if(TARGET ${eboot})
-    add_dependencies(${target} ${eboot})
+    add_dependencies(${target}.dep ${eboot})
   endif()
 endmacro(vita_create_vpk)
 ##################################################
```

While the velf thing is fixed and pkgj builds fine now, I can no longer properly create a dependency on self targets. For example I have this in a different project which used to work as expected:

https://github.com/cuevavirus/jav/blob/63957873f6402b2170e972174b53fd6e85a4a77c/kernel_module/CMakeLists.txt#L36-L52

but now produces `make[2]: *** No rule to make target 'kernel_module/jav_kernel.skprx', needed by 'kernel_module/jav_kernel.c'.  Stop.`

If I change `DEPENDS "${CMAKE_CURRENT_BINARY_DIR}/${SELF}"` to `DEPENDS "${SELF}"` it will build the first time but subsequent updates of the self target will not trigger the custom command to run again.

--- a/cmake_toolchain/vita.cmake
+++ b/cmake_toolchain/vita.cmake
@@ -152,39 +152,33 @@ macro(vita_create_stubs target-dir source config)
 
   ## ELF EXPORT target
   separate_arguments(VITA_LIBS_GEN_FLAGS)
-  add_custom_target(${target-dir}
-    ALL
-    COMMAND ${VITA_LIBS_GEN} ${VITA_LIBS_GEN_FLAGS} ${target_yml} ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
-    COMMAND make -C ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
-    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml
-    COMMENT "Building stubs ${target-dir}"
-  )
 
   set(stub_lib lib${target-lib}_stub.a)
   set(stub_weak_lib lib${target-lib}_stub_weak.a)
 
-  if(${CMAKE_GENERATOR} MATCHES ".*Ninja.*")
-    add_custom_target(${stub_lib}
-      ALL
-      COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_lib} ${stub_lib}
-      DEPENDS ${target-dir}
-    )
-    add_custom_target(${stub_weak_lib}
-      ALL
-      COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_weak_lib} ${stub_weak_lib}
-      DEPENDS ${target-dir}
-    )
-  else()
-    add_custom_command(TARGET ${target-dir}
-      POST_BUILD
-      COMMAND ${CMAKE_COMMAND} -E copy
-        ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_lib}
-        ${CMAKE_CURRENT_BINARY_DIR}/${stub_lib}
-      COMMAND ${CMAKE_COMMAND} -E copy
-        ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_weak_lib}
-        ${CMAKE_CURRENT_BINARY_DIR}/${stub_weak_lib}
-    )
-  endif()
+  add_custom_target(${target-dir}
+    ALL
+    COMMAND ${VITA_LIBS_GEN} ${VITA_LIBS_GEN_FLAGS} ${target_yml} ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
+    COMMAND make -C ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
+    COMMAND ${CMAKE_COMMAND} -E copy
+      ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_lib}
+      ${CMAKE_CURRENT_BINARY_DIR}/${stub_lib}
+    COMMAND ${CMAKE_COMMAND} -E copy
+      ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_weak_lib}
+      ${CMAKE_CURRENT_BINARY_DIR}/${stub_weak_lib}
+    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml
+    COMMENT "Building stubs ${target-dir}"
+  )
+
+  add_custom_target(${stub_lib}
+    ALL
+    DEPENDS ${target-dir}
+  )
+
+  add_custom_target(${stub_weak_lib}
+    ALL
+    DEPENDS ${target-dir}
+  )
 
   if(TARGET ${source})
     add_dependencies(${target-dir} ${source})

--- a/cmake_toolchain/vita.cmake
+++ b/cmake_toolchain/vita.cmake
@@ -75,10 +75,12 @@ macro(vita_create_self target source)
     COMMENT "Converting to Sony ELF ${sourcefile}.velf" VERBATIM
   )
 
+  set(self_outfile ${CMAKE_CURRENT_BINARY_DIR}/${target}.out)
+
   ## SELF command
   separate_arguments(VITA_MAKE_FSELF_FLAGS)
-  add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${target}
-    COMMAND ${VITA_MAKE_FSELF} ${VITA_MAKE_FSELF_FLAGS} ${CMAKE_CURRENT_BINARY_DIR}/${sourcefile}.velf ${CMAKE_CURRENT_BINARY_DIR}/${target}
+  add_custom_command(OUTPUT ${self_outfile}
+    COMMAND ${VITA_MAKE_FSELF} ${VITA_MAKE_FSELF_FLAGS} ${CMAKE_CURRENT_BINARY_DIR}/${sourcefile}.velf ${self_outfile}
     DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${sourcefile}.velf
     COMMENT "Creating SELF ${target}"
   )
@@ -86,8 +88,10 @@ macro(vita_create_self target source)
   ## SELF target
   add_custom_target(${target}
     ALL
-    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target}
+    DEPENDS ${self_outfile}
+    COMMAND ${CMAKE_COMMAND} -E copy ${self_outfile} ${target}
   )
+
   if(TARGET ${source})
     add_dependencies(${target} ${source})
   endif()
@@ -132,11 +136,15 @@ macro(vita_create_stubs target-dir source config)
   endif()
   get_filename_component(sourcefile ${sourcepath} NAME)
 
+  set(target_yml ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml)
+
   ## ELF EXPORT command
   separate_arguments(VITA_ELF_EXPORT_FLAGS)
   get_filename_component(fconfig ${config} ABSOLUTE)
-  add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml
-    COMMAND ${VITA_ELF_EXPORT} ${kind} ${VITA_ELF_EXPORT_FLAGS} ${sourcepath} ${fconfig} ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml
+  file(READ ${fconfig} fconfig_content)
+  string(REGEX REPLACE ":.+" "" target-lib ${fconfig_content})
+  add_custom_command(OUTPUT ${target_yml}
+    COMMAND ${VITA_ELF_EXPORT} ${kind} ${VITA_ELF_EXPORT_FLAGS} ${sourcepath} ${fconfig} ${target_yml}
     DEPENDS ${sourcepath}
     DEPENDS ${fconfig}
     COMMENT "Generating imports YAML for ${sourcefile}"
@@ -146,11 +154,38 @@ macro(vita_create_stubs target-dir source config)
   separate_arguments(VITA_LIBS_GEN_FLAGS)
   add_custom_target(${target-dir}
     ALL
-    COMMAND ${VITA_LIBS_GEN} ${VITA_LIBS_GEN_FLAGS} ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
+    COMMAND ${VITA_LIBS_GEN} ${VITA_LIBS_GEN_FLAGS} ${target_yml} ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
     COMMAND make -C ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}
     DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}.yml
     COMMENT "Building stubs ${target-dir}"
   )
+
+  set(stub_lib lib${target-lib}_stub.a)
+  set(stub_weak_lib lib${target-lib}_stub_weak.a)
+
+  if(${CMAKE_GENERATOR} MATCHES ".*Ninja.*")
+    add_custom_target(${stub_lib}
+      ALL
+      COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_lib} ${stub_lib}
+      DEPENDS ${target-dir}
+    )
+    add_custom_target(${stub_weak_lib}
+      ALL
+      COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_weak_lib} ${stub_weak_lib}
+      DEPENDS ${target-dir}
+    )
+  else()
+    add_custom_command(TARGET ${target-dir}
+      POST_BUILD
+      COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_lib}
+        ${CMAKE_CURRENT_BINARY_DIR}/${stub_lib}
+      COMMAND ${CMAKE_COMMAND} -E copy
+        ${CMAKE_CURRENT_BINARY_DIR}/${target-dir}/${stub_weak_lib}
+        ${CMAKE_CURRENT_BINARY_DIR}/${stub_weak_lib}
+    )
+  endif()
+
   if(TARGET ${source})
     add_dependencies(${target-dir} ${source})
   endif()
@@ -228,10 +263,12 @@ macro(vita_create_vpk target titleid eboot)
     COMMENT "Generating param.sfo for ${target}"
   )
 
+  set(vpk_outfile ${CMAKE_CURRENT_BINARY_DIR}/${target}.out)
+
   ## VPK command
   separate_arguments(VITA_PACK_VPK_FLAGS)
-  add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${target}
-    COMMAND ${VITA_PACK_VPK} ${VITA_PACK_VPK_FLAGS} -s ${CMAKE_CURRENT_BINARY_DIR}/${target}_param.sfo -b ${sourcepath} ${CMAKE_CURRENT_BINARY_DIR}/${target}
+  add_custom_command(OUTPUT ${vpk_outfile}
+    COMMAND ${VITA_PACK_VPK} ${VITA_PACK_VPK_FLAGS} -s ${CMAKE_CURRENT_BINARY_DIR}/${target}_param.sfo -b ${sourcepath} ${vpk_outfile}
     DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target}_param.sfo
     DEPENDS ${sourcepath}
     DEPENDS ${resources}
@@ -241,8 +278,10 @@ macro(vita_create_vpk target titleid eboot)
   ## VPK target
   add_custom_target(${target}
     ALL
-    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${target}
+    DEPENDS ${vpk_outfile}
+    COMMAND ${CMAKE_COMMAND} -E copy ${vpk_outfile} ${target}
   )
+
   if(TARGET ${eboot})
     add_dependencies(${target} ${eboot})
   endif()

